agent/engine/dockerstate/json.go

```diff
@@ -55,47 +55,50 @@ func (state *DockerTaskEngineState) UnmarshalJSON(data []byte) error {
 	if err != nil {
 		return err
 	}
-	// reset it by just creating a new one and swapping shortly.
-	// This also means we don't have to lock for the remainder of this function
-	// because we are the only ones with a reference to clean
-	clean := newDockerTaskEngineState()
 
+	// run precheck; shake out all the errors before resetting state.
+	precheckState := newDockerTaskEngineState()
 	for _, task := range saved.Tasks {
-		clean.AddTask(task)
-	}
-	// add image states
-	for _, imageState := range saved.ImageStates {
-		clean.AddImageState(imageState)
+		precheckState.AddTask(task)
 	}
 	for id, container := range saved.IdToContainer {
 		taskArn, ok := saved.IdToTask[id]
 		if !ok {
 			return errors.New("Could not unmarshal state; incomplete save. There was no task for docker id " + id)
 		}
-		task, ok := clean.TaskByArn(taskArn)
+		task, ok := precheckState.TaskByArn(taskArn)
 		if !ok {
 			return errors.New("Could not unmarshal state; incomplete save. There was no task for arn " + taskArn)
 		}
-
 		// The container.Container pointers *must* match the task's container
 		// pointers for things to operate correctly; update them here
-		taskContainer, ok := task.ContainerByName(container.Container.Name)
+		_, ok = task.ContainerByName(container.Container.Name)
 		if !ok {
 			return errors.New("Could not resolve a container into a task based on name: " + task.String() + " -- " + container.String())
 		}
-		container.Container = taskContainer
-		//pointer matching now; everyone happy
-		clean.AddContainer(container, task)
 	}
 
+	// reset state and safely populate with saved
+	state.Reset()
+	for _, task := range saved.Tasks {
+		state.AddTask(task)
+	}
```

Comment:
Can state still be changed between the different calls i.e state could be changed between `AddTask` and `AddContainer`?

Reply:
Yes, it can still be changed if its lock is acquired between these calls.  As far as I can tell,  unmarshalling json state into the docker state struct is only done when we start up agent from a backup.  Where are we worried about contention for this lock given the context of unmarshalling from json?  
I believe this issue was at least one big driver for copy (`*state = *clean`) in the first place.  The issue is that `go vet` is unequivocal: thou shalt not ever copy a mutex.
Any thoughts on how else to get around that would be great.

Reply:
I was under the impression that this only happens at startup -- which means this should be safe.